dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramEventSupplier.java

```diff
@@ -0,0 +1,424 @@
+        Environment environment )
+    {
+        super( jdbcTemplate, environment );
+        programsCache = cacheProvider.createProgramCache();
+    }
+
+    public Map<String, Program> get( ImportOptions importOptions, List<Event> eventList )
+    {
+        boolean requiresReload = false;
+
+        // Do not use cache if {@code skipCache} is true or if running as test
+
+        if ( importOptions.isSkipCache() )
+        {
+            programsCache.invalidateAll();
+        }
+        Map<String, Program> programMap = programsCache.get( PROGRAM_CACHE_KEY ).orElse( null );
+
+        if ( requiresCacheReload( eventList, programMap ) )
+        {
+            programsCache.invalidateAll();
+            requiresReload = true;
+        }
+
+        if ( requiresReload || programMap == null )
+        {
+            // Load all programs
+
+            programMap = loadPrograms( importOptions.getIdSchemes() );
+
+            if ( !MapUtils.isEmpty( programMap ) )
+            {
+                aggregateProgramAndAclData( programMap,
+                    loadProgramStageDataElementSets() );
+
+                programsCache.put( PROGRAM_CACHE_KEY, programMap );
+            }
+            else
+            {
+                programsCache.put( PROGRAM_CACHE_KEY, new HashMap<>() );
+            }
+
+        }
+
+        return programMap;
+    }
+
+    private void aggregateProgramAndAclData( Map<String, Program> programMap,
+        Map<Long, DataElementSets> dataElementSetsMap )
+    {
+        for ( Program program : programMap.values() )
+        {
+            for ( ProgramStage programStage : program.getProgramStages() )
+            {
+                DataElementSets dataElementSets = dataElementSetsMap.get( programStage.getId() );
+                if ( dataElementSets != null )
+                {
+                    Stream<ProgramStageDataElement> compulsoryProgramStageDataElementsStream = getOrEmpty(
+                        dataElementSets.getCompulsoryDataElements() )
+                            .map( de -> new ProgramStageDataElement( programStage, de, true ) );
+
+                    Stream<ProgramStageDataElement> nonCompulsoryProgramStageDataElementsStream = getOrEmpty(
+                        dataElementSets.getNonCompulsoryDataElements() )
+                            .map( de -> new ProgramStageDataElement( programStage, de, false ) );
+
+                    Set<ProgramStageDataElement> allProgramStageDataElements = Stream
+                        .concat( compulsoryProgramStageDataElementsStream, nonCompulsoryProgramStageDataElementsStream )
+                        .collect( Collectors.toSet() );
+
+                    programStage.setProgramStageDataElements( allProgramStageDataElements );
+                }
+            }
+        }
+    }
+
+    private Stream<DataElement> getOrEmpty( Set<DataElement> dataElementSet )
+    {
+        return Optional.ofNullable( dataElementSet )
```

Comment:
In my opinion all these Optional.ofNullable are not adding any value to the code and they are making the flow harder to follow.
In here you are sure that a dataElementSet will never be null.

Reply:
This is a sonar suggestion